File: pangath_properties/events/purchase_invoice.py
import frappe

def after_insert_purchase_invoice(doc, method):
    if doc.inter_company_invoice_reference:
        sales_invoice_name = doc.inter_company_invoice_reference
        sales_invoice = frappe.get_doc("Sales Invoice", sales_invoice_name)
        sales_invoice.custom_inter_company_purchase_invoice_reference = doc.name
        sales_invoice.save(ignore_permissions=True)


@frappe.whitelist()
def get_units_property_items(data):
    data = frappe.parse_json(data)
    result = []

    for row in data:
        qty = row.get('qty') or 0
        rate = row.get('rate') or 0
        entry_for = row.get('entry_for')

        if row.get("item_code"):
            item = frappe.db.get_value(
                "Item", row.get("item_code"),
                ["item_name", "stock_uom"],
                as_dict=True
            )
        else:
            item = {"item_name": None, "stock_uom": None}

        if entry_for == 'Property':
            unit_list = frappe.db.get_list(
                'Unit',
                filters={'property': row.get('property')},
                fields=['name', 'property']
            )
            num_units = len(unit_list)
            per_qty = qty / num_units if num_units else qty

            if num_units > 0:
                for unit in unit_list:
                    entry = row.copy()
                    entry['property'] = unit.get('property')
                    entry['unit'] = unit.get('name')
                    entry['qty'] = per_qty
                    entry['amount'] = per_qty * rate
                    entry['item_name'] = item.get("item_name")
                    entry['uom'] = item.get("stock_uom")
                    result.append(entry)
            else:
                row['item_name'] = item.get("item_name")
                row['uom'] = item.get("stock_uom")
                result.append(row)

        elif entry_for == 'Unit':
            list_unit = frappe.db.get_list(
                'Unit',
                filters={'is_group': 0, 'parent_unit': row.get('unit')},
                fields=['name']
            )
            num_units = len(list_unit)
            per_qty = qty / num_units if num_units else qty

            if num_units > 0:
                for unit in list_unit:
                    entry = row.copy()
                    entry['unit'] = unit.get('name')
                    entry['qty'] = per_qty
                    entry['amount'] = per_qty * rate
                    entry['item_name'] = item.get("item_name")
                    entry['uom'] = item.get("stock_uom")
                    result.append(entry)
            else:
                row['item_name'] = item.get("item_name")
                row['uom'] = item.get("stock_uom")
                result.append(row)

        else:
            row['item_name'] = item.get("item_name")
            row['uom'] = item.get("stock_uom")
            result.append(row)

    return result

File: pangath_properties/public/js/purchase_invoice.js
frappe.ui.form.on("Purchase Invoice", {
    refresh: function (frm) {
        setTimeout(() => {
            frm.set_query('customer', function (doc) {
                return { "filters": [] };
            });
        }, 2000);

        frm.add_custom_button('Insert Records', function () {
            let d = new frappe.ui.Dialog({
                title: 'Insert Records',
                fields: [
                    {
                        label: 'Items Table',
                        fieldname: 'items_table',
                        fieldtype: 'Table',
                        in_place_edit: true,
                        fields: [
                            {
                                fieldtype: 'Link',
                                label: 'Item Code',
                                fieldname: 'item_code',
                                options: 'Item',
                                reqd: 1,
                                in_list_view: 1,
                                columns: 2
                            },
                            {
                                fieldtype: 'Select',
                                label: 'Entry For',
                                fieldname: 'entry_for',
                                options: ['Unit', 'Property'],
                                in_list_view: 1,
                                columns: 2
                            },
                            {
                                fieldtype: 'Link',
                                label: 'Unit',
                                fieldname: 'unit',
                                options: 'Unit',
                                in_list_view: 1,
                                columns: 1,
                                read_only_depends_on: 'eval:doc.entry_for=="Property"'
                            },
                            {
                                fieldtype: 'Link',
                                label: 'Property',
                                fieldname: 'property',
                                options: 'Property',
                                in_list_view: 1,
                                columns: 2,
                                read_only_depends_on: 'eval:doc.entry_for=="Unit"'
                            },
                            {
                                fieldtype: 'Float',
                                label: 'Qty',
                                fieldname: 'qty',
                                columns: 1,
                                default: 1
                            },
                            {
                                fieldtype: 'Currency',
                                label: 'Rate',
                                fieldname: 'rate',
                                in_list_view: 1,
                                columns: 2
                            }
                        ]
                    }
                ],

                primary_action(values) {
                    frm.clear_table('items');
                    frappe.call({
                        method: 'pangath_properties.events.purchase_invoice.get_units_property_items',
                        args: {
                            data: values.items_table
                        },
                        callback: function (r) {
                            if (r.message) {
                                r.message.forEach(row => {
                                    frm.add_child('items', {
                                        item_code: row.item_code,
                                        qty: 1,
                                        item_name: row.item_name,
                                        uom: row.uom,
                                        rate: row.qty * row.rate,
                                        amount: row.qty * row.rate,
                                        property: row.property,
                                        unit: row.unit
                                    });
                                });
                                frm.refresh_field('items');
                            }
                        }
                    });
                    d.hide();
                }
            });

            d.show();
        });
    }
});
